components/movementsensor/wheeledodometry/wheeled_odometry.go

```diff
@@ -0,0 +1,341 @@
+// linear velocity, and angular velocity of the wheeled base.
+// The estimations in this function are based on the math outlined in this article:
+// https://stuff.mit.edu/afs/athena/course/6/6.186/OldFiles/2005/doc/odomtutorial/odomtutorial.pdf
+func (o *odometry) trackPosition(ctx context.Context) {
+	o.activeBackgroundWorkers.Add(1)
+	utils.PanicCapturingGo(func() {
+		defer o.activeBackgroundWorkers.Done()
+		ticker := time.NewTicker(time.Duration(o.timeIntervalMSecs) * time.Millisecond)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			// Sleep until it's time to update the position again.
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+			}
+
+			// always use the first pair until more than one pair of motors is supported in this model
+			leftMove, err := o.motors[0].left.IsMoving(ctx)
+			if err != nil {
+				o.logger.Error(err)
+				return
+			}
+			rightMove, err := o.motors[0].right.IsMoving(ctx)
+			if err != nil {
+				o.logger.Error(err)
+				return
+			}
+			if !leftMove && !rightMove {
+				// neither motors are moving, so the odometry values do not need to be updated
+				o.linearVelocity.Y = 0
+				o.angularVelocity.Z = 0
+				continue
+			}
+
+			// use GetInParallel to ensure the left and right motors are polled at the same time
+			positionFuncs := func() []rdkutils.FloatFunc {
+				fs := []rdkutils.FloatFunc{}
+
+				o.mu.Lock()
+				defer o.mu.Unlock()
+				// always use the first pair until more than one pair of motors is supported in this model
+				fs = append(fs, func(ctx context.Context) (float64, error) { return o.motors[0].left.Position(ctx, nil) })
+				fs = append(fs, func(ctx context.Context) (float64, error) { return o.motors[0].right.Position(ctx, nil) })
+
+				return fs
+			}
+
+			_, positions, err := rdkutils.GetInParallel(ctx, positionFuncs())
+			if err != nil {
+				o.logger.Error(err)
+				continue
+			}
+
+			// current position of the left and right motors in revolutions
+			if len(positions) != len(o.motors)*2 {
+				o.logger.Error("error getting both motor positions, trying again")
+				continue
+			}
+			left := positions[0]
+			right := positions[1]
+
+			// difference in the left and right motors since the last iteration, in mm
+			leftDist := (left - o.lastLeftPos) * o.wheelCircumference
+			rightDist := (right - o.lastRightPos) * o.wheelCircumference
+
+			// update lastLeftPos and lastRightPos to be the current position in mm
+			o.lastLeftPos = left
+			o.lastRightPos = right
+
+			// linear and angular distance the center point has traveled
+			centerDist := (leftDist + rightDist) / 2
+			centerAngle := (rightDist - leftDist) / o.baseWidth
+
+			// update the position and orientation values accordingly
+			o.mu.Lock()
+			o.orientation.Yaw += centerAngle
+
+			o.orientation.Yaw = math.Mod(o.orientation.Yaw, oneTurn)
+			o.orientation.Yaw = math.Mod(o.orientation.Yaw+oneTurn, oneTurn)
```

Comment:
The idea here is that the first `math.Mod` puts us between `-oneTurn` and `+oneTurn`, and the second one forces us to be positive, yeah? Put that in a comment, because it's not clear otherwise.

...given that we're making the small angle assumption, the yaw should never change by more than 2pi at a time, right? I wonder if we don't even need line 328.

Reply:
will definitely add a comment about it 👍 

hmm yeah maybe not. but technically a user can put a large time interval so I still think even if that causes it to be inaccurate, it should probably still be limited to that range?